panic: Sensitive values ​​in module sources cause a panic

### Terraform Version

```shell
Terraform v1.15.5
on linux_amd64
```

### Terraform Configuration Files

```terraform
variable "sensitive" {
  default   = "./module"
  sensitive = true
  const     = true
}

module "main" {
  source = var.sensitive
}
```


### Debug Output

```
2026-06-07T22:43:04.558+0900 [INFO]  Terraform version: 1.15.5
2026-06-07T22:43:04.559+0900 [DEBUG] using github.com/hashicorp/go-tfe v1.105.0
2026-06-07T22:43:04.559+0900 [DEBUG] using github.com/hashicorp/hcl/v2 v2.24.0
2026-06-07T22:43:04.559+0900 [DEBUG] using github.com/hashicorp/terraform-svchost v0.2.1
2026-06-07T22:43:04.559+0900 [DEBUG] using github.com/zclconf/go-cty v1.18.0
2026-06-07T22:43:04.559+0900 [INFO]  Go runtime version: go1.25.10
2026-06-07T22:43:04.559+0900 [INFO]  CLI args: []string{"terraform", "init"}
2026-06-07T22:43:04.559+0900 [DEBUG] Attempting to open CLI config file: /home/wata727/.terraformrc
2026-06-07T22:43:04.559+0900 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2026-06-07T22:43:04.559+0900 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2026-06-07T22:43:04.559+0900 [DEBUG] ignoring non-existing provider search directory /home/wata727/.terraform.d/plugins
2026-06-07T22:43:04.559+0900 [DEBUG] ignoring non-existing provider search directory /home/wata727/.local/share/terraform/plugins
2026-06-07T22:43:04.559+0900 [DEBUG] ignoring non-existing provider search directory /usr/local/share/terraform/plugins
2026-06-07T22:43:04.559+0900 [DEBUG] ignoring non-existing provider search directory /usr/share/terraform/plugins
2026-06-07T22:43:04.559+0900 [DEBUG] ignoring non-existing provider search directory /var/lib/snapd/desktop/terraform/plugins
2026-06-07T22:43:04.559+0900 [INFO]  CLI command args: []string{"init"}
Initializing modules...
2026-06-07T22:43:04.560+0900 [DEBUG] ReferenceTransformer: "var.sensitive" references: []
2026-06-07T22:43:04.560+0900 [DEBUG] ReferenceTransformer: "module.main" references: [var.sensitive]
2026-06-07T22:43:04.560+0900 [DEBUG] Starting graph walk: walkInit
2026-06-07T22:43:04.561+0900 [ERROR] vertex "module.main" panicked

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

panic: value is marked, so must be unmarked first
goroutine 70 [running]:
runtime/debug.Stack()
        runtime/debug/stack.go:26 +0x5e
github.com/hashicorp/terraform/internal/logging.PanicHandler()
        github.com/hashicorp/terraform/internal/logging/panic.go:84 +0x16a
panic({0x3bfc8e0?, 0x4da6e70?})
        runtime/panic.go:783 +0x132
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1.1()
        github.com/hashicorp/terraform/internal/terraform/graph.go:59 +0x4e5
panic({0x3bfc8e0?, 0x4da6e70?})
        runtime/panic.go:783 +0x132
github.com/zclconf/go-cty/cty.Value.assertUnmarked(...)
        github.com/zclconf/go-cty@v1.18.0/cty/marks.go:191
github.com/zclconf/go-cty/cty.Value.AsString({{{0x4e11b98?, 0x4d9ad90?}}, {0x4086820?, 0xc000010b58?}})
        github.com/zclconf/go-cty@v1.18.0/cty/value_ops.go:1456 +0x47
github.com/hashicorp/terraform/internal/terraform.evalSource({0x4e124d8, 0xc000152b40}, 0x0, {0x4e4e6a0, 0xc00042e120})
        github.com/hashicorp/terraform/internal/terraform/node_module_install.go:217 +0x539
github.com/hashicorp/terraform/internal/terraform.(*nodeInstallModule).Execute(0xc00070b590, {0x4e4e6a0, 0xc00042e120}, 0x58?)
        github.com/hashicorp/terraform/internal/terraform/node_module_install.go:94 +0x2b1
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc000980b00, {0x4e4e6a0, 0xc00042e120}, {0x7beb469599f0, 0xc00070b590})
        github.com/hashicorp/terraform/internal/terraform/graph_walk_context.go:170 +0xaf
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x4151480, 0xc00070b590})
        github.com/hashicorp/terraform/internal/terraform/graph.go:143 +0x755
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc000153680, {0x4151480, 0xc00070b590}, 0xc0007064c0)
        github.com/hashicorp/terraform/internal/dag/walk.go:393 +0x2d1
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 1
        github.com/hashicorp/terraform/internal/dag/walk.go:316 +0xf13
```


### Expected Behavior

It's difficult to define the expected behavior in this case, but conservatively, it should be treated as an error.

### Actual Behavior

It will cause panic.

### Steps to Reproduce

1. `terraform init`

### Additional Context

`value.AsString()` will cause a panic for marked values.

https://github.com/hashicorp/terraform/blob/d038b9e6cd8636e97a6844e45e8a0157a50ef8df/internal/terraform/node_module_install.go#L201-L217
https://github.com/zclconf/go-cty/blob/v1.18.0/cty/value_ops.go#L1456

Just to be clear, this issue did not occur in actual production code. It was discovered during white-box testing when [supporting dynamic module sources in TFLint](https://github.com/terraform-linters/tflint/pull/2521), and such cases are likely to be rare in practice.

### References

- https://github.com/hashicorp/terraform/pull/38217
- https://github.com/hashicorp/terraform/pull/38628

### Generative AI / LLM assisted development?

_No response_

--- a/.changes/v1.15/BUG FIXES-20260610-100345.yaml
+++ b/.changes/v1.15/BUG FIXES-20260610-100345.yaml
@@ -1,0 +1,5 @@
+kind: BUG FIXES
+body: Fix two module installation edge cases with `null` and sensitive/ephemeral module sources
+time: 2026-06-10T10:03:45.715338+02:00
+custom:
+    Issue: "38704"

--- a/internal/terraform/context_init_test.go
+++ b/internal/terraform/context_init_test.go
@@ -58,6 +58,9 @@ func TestInit(t *testing.T) {
 		// mc -> module calls
 		expectDiags           func(m *configs.Module, mc map[string]*configs.Module) tfdiags.Diagnostics
 		expectLoadModuleCalls []*configs.ModuleRequest
+		// Set to true to emulate module installation continuing even in the presence
+		// of parsing errors
+		ignoreParsingErrors bool
 	}{
 		"empty config": {
 			module: map[string]string{"main.tf": ``},
@@ -847,20 +850,174 @@ module "local" {
 				})
 			},
 		},
+
+		"source from const sensitive variable": {
+			module: map[string]string{
+				"main.tf": `
+variable "sourc" {
+  const     = true
+  sensitive = true
+  default   = "./modules/local"
+}
+module "local" {
+  source  = var.sourc
+}
+`,
+			},
+			ignoreParsingErrors: true,
+			expectDiags: func(m *configs.Module, mc map[string]*configs.Module) tfdiags.Diagnostics {
+				return tfdiags.Diagnostics{}.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Unknown module source",
+					Detail:   "The module source cannot be derived from a sensitive or ephemeral value.",
+					Subject: &hcl.Range{
+						Filename: filepath.Join(m.SourceDir, "main.tf"),
+						Start:    hcl.Pos{Line: 8, Column: 13, Byte: 121},
+						End:      hcl.Pos{Line: 8, Column: 22, Byte: 130},
+					},
+				})
+			},
+		},
+
+		"source from const ephemeral variable": {
+			module: map[string]string{
+				"main.tf": `
+variable "sourc" {
+  const     = true
+  ephemeral = true
+  default   = "./modules/local"
+}
+module "local" {
+  source  = var.sourc
+}
+`,
+			},
+			ignoreParsingErrors: true,
+			expectDiags: func(m *configs.Module, mc map[string]*configs.Module) tfdiags.Diagnostics {
+				return tfdiags.Diagnostics{}.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Unknown module source",
+					Detail:   "The module source cannot be derived from a sensitive or ephemeral value.",
+					Subject: &hcl.Range{
+						Filename: filepath.Join(m.SourceDir, "main.tf"),
+						Start:    hcl.Pos{Line: 8, Column: 13, Byte: 121},
+						End:      hcl.Pos{Line: 8, Column: 22, Byte: 130},
+					},
+				})
+			},
+		},
+
+		"source from const deprecated variable": {
+			module: map[string]string{
+				"main.tf": `
+variable "sourc" {
+  deprecated = true
+  const      = true
+  default    = "./modules/local"
+}
+module "local" {
+  source  = var.sourc
+}
+`,
+			},
+			ignoreParsingErrors: true,
+			expectLoadModuleCalls: []*configs.ModuleRequest{{
+				SourceAddr: mustModuleSource(t, "./modules/local"),
+			}},
+		},
+
+		"version from const sensitive variable": {
+			module: map[string]string{
+				"main.tf": `
+variable "ver" {
+  const     = true
+  sensitive = true
+  default   = "1.0.0"
+}
+module "local" {
+  source  = "./modules/local"
+  version = var.ver
+}
+`,
+			},
+			ignoreParsingErrors: true,
+			expectDiags: func(m *configs.Module, mc map[string]*configs.Module) tfdiags.Diagnostics {
+				return tfdiags.Diagnostics{}.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Unknown module version",
+					Detail:   "The module version cannot be derived from a sensitive or ephemeral value.",
+					Subject: &hcl.Range{
+						Filename: filepath.Join(m.SourceDir, "main.tf"),
+						Start:    hcl.Pos{Line: 9, Column: 13, Byte: 139},
+						End:      hcl.Pos{Line: 9, Column: 20, Byte: 146},
+					},
+				})
+			},
+		},
+
+		"version from const ephemeral variable": {
+			module: map[string]string{
+				"main.tf": `
+variable "ver" {
+  const     = true
+  ephemeral = true
+  default   = "1.0.0"
+}
+module "local" {
+  source  = "./modules/local"
+  version = var.ver
+}
+`,
+			},
+			ignoreParsingErrors: true,
+			expectDiags: func(m *configs.Module, mc map[string]*configs.Module) tfdiags.Diagnostics {
+				return tfdiags.Diagnostics{}.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Unknown module version",
+					Detail:   "The module version cannot be derived from a sensitive or ephemeral value.",
+					Subject: &hcl.Range{
+						Filename: filepath.Join(m.SourceDir, "main.tf"),
+						Start:    hcl.Pos{Line: 9, Column: 13, Byte: 139},
+						End:      hcl.Pos{Line: 9, Column: 20, Byte: 146},
+					},
+				})
+			},
+		},
+
+		"version from const deprecated variable": {
+			module: map[string]string{
+				"main.tf": `
+variable "ver" {
+  deprecated = true
+  const      = true
+  default   = "1.0.0"
+}
+module "local" {
+  source  = "terraform-aws-modules/vpc/aws"
+  version = var.ver
+}
+`,
+			},
+			ignoreParsingErrors: true,
+			expectLoadModuleCalls: []*configs.ModuleRequest{{
+				SourceAddr:        mustModuleSource(t, "terraform-aws-modules/vpc/aws"),
+				VersionConstraint: mustVersionContraint(t, "1.0.0"),
+			}},
+		},
 	} {
 		t.Run(name, func(t *testing.T) {
-			m := testRootModuleInline(t, tc.module)
+			m := testRootModuleInline(t, tc.module, tc.ignoreParsingErrors)
 
 			ctx := testContext2(t, &ContextOpts{
 				Parallelism: 1,
 			})
 			moduleWalker := MockModuleWalker{
-				DefaultModule: testRootModuleInline(t, map[string]string{"main.tf": `// empty`}),
+				DefaultModule: testRootModuleInline(t, map[string]string{"main.tf": `// empty`}, false),
 			}
 			mockedModules := make(map[string]*configs.Module)
 			if tc.mockedLoadModuleCalls != nil {
 				for k, v := range tc.mockedLoadModuleCalls {
-					mockedModules[k] = testRootModuleInline(t, v)
+					mockedModules[k] = testRootModuleInline(t, v, false)
 				}
 				moduleWalker.MockModuleCalls(t, mockedModules)
 			}

--- a/internal/terraform/context_init_test.go
+++ b/internal/terraform/context_init_test.go
@@ -821,6 +821,32 @@ module "local" {
 				SourceAddr: mustModuleSource(t, "./modules/local"),
 			}},
 		},
+
+		"source from const variable set to null": {
+			module: map[string]string{
+				"main.tf": `
+variable "sourc" {
+  const   = true
+  default = null
+}
+module "local" {
+  source  = var.sourc
+}
+`,
+			},
+			expectDiags: func(m *configs.Module, mc map[string]*configs.Module) tfdiags.Diagnostics {
+				return tfdiags.Diagnostics{}.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Unsuitable module source",
+					Detail:   `Unsuitable value: null value is not allowed.`,
+					Subject: &hcl.Range{
+						Filename: filepath.Join(m.SourceDir, "main.tf"),
+						Start:    hcl.Pos{Line: 7, Column: 13, Byte: 85},
+						End:      hcl.Pos{Line: 7, Column: 22, Byte: 94},
+					},
+				})
+			},
+		},
 	} {
 		t.Run(name, func(t *testing.T) {
 			m := testRootModuleInline(t, tc.module)

--- a/internal/terraform/node_module_install.go
+++ b/internal/terraform/node_module_install.go
@@ -14,6 +14,7 @@ import (
 	"github.com/hashicorp/terraform/internal/configs"
 	"github.com/hashicorp/terraform/internal/getmodules/moduleaddrs"
 	"github.com/hashicorp/terraform/internal/lang/langrefs"
+	"github.com/hashicorp/terraform/internal/lang/marks"
 	"github.com/hashicorp/terraform/internal/tfdiags"
 )
 
@@ -224,6 +225,20 @@ func evalSource(sourceExpr hcl.Expression, hasVersion bool, ctx EvalContext) (ad
 		return nil, "", diags
 	}
 
+	if marks.Has(value, marks.Sensitive) || marks.Has(value, marks.Ephemeral) {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Unknown module source",
+			Detail:   "The module source cannot be derived from a sensitive or ephemeral value.",
+			Subject:  sourceExpr.Range().Ptr(),
+		})
+		return nil, "", diags
+	}
+
+	// Strip any remaining marks (such as deprecation marks, which are allowed)
+	// so that AsString cannot panic.
+	value, _ = value.Unmark()
+
 	rawSource := value.AsString()
 	if hasVersion {
 		addr, err = moduleaddrs.ParseModuleSourceRegistry(rawSource)
@@ -330,6 +345,20 @@ func evalVersionConstraint(versionExpr hcl.Expression, ctx EvalContext) (configs
 		})
 		return ret, diags
 	}
+
+	if marks.Has(value, marks.Sensitive) || marks.Has(value, marks.Ephemeral) {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Unknown module version",
+			Detail:   "The module version cannot be derived from a sensitive or ephemeral value.",
+			Subject:  versionExpr.Range().Ptr(),
+		})
+		return ret, diags
+	}
+
+	// Strip any remaining marks (such as deprecation marks, which are allowed)
+	// so that AsString cannot panic.
+	value, _ = value.Unmark()
 
 	constraintStr := value.AsString()
 	constraints, err := version.NewConstraint(constraintStr)

--- a/internal/terraform/node_module_install.go
+++ b/internal/terraform/node_module_install.go
@@ -204,6 +204,16 @@ func evalSource(sourceExpr hcl.Expression, hasVersion bool, ctx EvalContext) (ad
 		return nil, "", diags
 	}
 
+	if value.IsNull() {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Unsuitable module source",
+			Detail:   `Unsuitable value: null value is not allowed.`,
+			Subject:  sourceExpr.Range().Ptr(),
+		})
+		return nil, "", diags
+	}
+
 	if !value.IsWhollyKnown() {
 		diags = diags.Append(&hcl.Diagnostic{
 			Severity: hcl.DiagError,

--- a/internal/terraform/terraform_test.go
+++ b/internal/terraform/terraform_test.go
@@ -194,7 +194,7 @@ func testModuleInlineWithVars(t testing.TB, sources map[string]string, vars Inpu
 	return config
 }
 
-func testRootModuleInline(t testing.TB, sources map[string]string) *configs.Module {
+func testRootModuleInline(t testing.TB, sources map[string]string, ignoreParsingErrors bool) *configs.Module {
 	t.Helper()
 
 	cfgPath, err := filepath.EvalSymlinks(t.TempDir())
@@ -230,7 +230,7 @@ func testRootModuleInline(t testing.TB, sources map[string]string) *configs.Modu
 	loader.AllowLanguageExperiments(true)
 
 	mod, diags := loader.Parser().LoadConfigDir(cfgPath)
-	if diags.HasErrors() {
+	if diags.HasErrors() && !ignoreParsingErrors {
 		t.Fatal(diags.Error())
 	}
 

